File: typespec/employer/posts.go
package employer

import "github.com/vetchium/vetchium/typespec/common"

type AddEmployerPostRequest struct {
	Content string          `json:"content" validate:"required,min=1,max=4096"`
	TagIDs  []common.VTagID `json:"tag_ids" validate:"max=3"`
}

type AddEmployerPostResponse struct {
	PostID string `json:"post_id"`
}

type UpdateEmployerPostRequest struct {
	PostID  string          `json:"post_id"`
	Content string          `json:"content" validate:"required,min=1,max=4096"`
	TagIDs  []common.VTagID `json:"tags"    validate:"max=3"`
}

type DeleteEmployerPostRequest struct {
	PostID string `json:"post_id" validate:"required"`
}

type ListEmployerPostsRequest struct {
	PaginationKey string `json:"pagination_key"`
	Limit         int    `json:"limit"          validate:"min=0,max=40"`
}

type ListEmployerPostsResponse struct {
	Posts         []common.EmployerPost `json:"posts"`
	PaginationKey string                `json:"pagination_key"`
}

type GetEmployerPostRequest struct {
	PostID string `json:"post_id" validate:"required"`
}

File: typespec/employer/posts.ts
import { EmployerPost } from "../common/posts";

export class AddEmployerPostRequest {
  content: string = "";
  tag_ids: string[] = [];

  IsValid(): boolean {
    return (
      this.content.length > 0 &&
      this.content.length <= 4096 &&
      this.tag_ids.length > 0 &&
      this.tag_ids.length <= 3 &&
      this.tag_ids.every((tag) => tag.length > 0 && tag.length <= 64)
    );
  }
}

export interface AddEmployerPostResponse {
  post_id: string;
}

export class UpdateEmployerPostRequest {
  post_id: string = "";
  content: string = "";
  tags: string[] = [];

  IsValid(): boolean {
    return (
      this.content.length > 0 &&
      this.content.length <= 4096 &&
      this.tags.length > 0 &&
      this.tags.length <= 3
    );
  }
}

export interface DeleteEmployerPostRequest {
  post_id: string;
}

export class ListEmployerPostsRequest {
  pagination_key?: string;
  limit?: number;

  IsValid(): boolean {
    return this.limit !== undefined && this.limit >= 0 && this.limit <= 40;
  }
}

export interface ListEmployerPostsResponse {
  posts: EmployerPost[];
  pagination_key: string;
}

export class GetEmployerPostRequest {
  post_id: string = "";
}
